### PR TITLE
Feature/dismantle mmfiles

### PR DESCRIPTION
**Pull request by jsteemann**
### Scope & Purpose

Dismantle MMFiles storage engine in devel.
Enterprise companion PR: https://github.com/arangodb/enterprise/pull/423
Oskar PR: https://github.com/arangodb/oskar/pull/233

Documentation is still to do.

- [x] Strictly *new functionality* (i.e. a new feature / new option, no need for porting)
- [x] The behavior in this PR can be (and was) *manually tested* (support / qa / customers can test it)
- [x] The behavior change can be verified via automatic tests

### Testing & Verification

This change is already covered by existing tests, such as *scripts/unittest all*.

### CODE REVIEW

**Comment by jsteemann** [comment]
http://172.16.10.101:8080/view/PR/job/arangodb-matrix-pr/9242/

**Comment by jsteemann** [comment]
http://172.16.10.101:8080/view/PR/job/arangodb-matrix-pr/9243/

File: arangod/RocksDBEngine/RocksDBCollection.cpp [hunk 1]
```diff
@@ -776,7 +776,7 @@ bool RocksDBCollection::lookupRevision(transaction::Methods* trx, VPackSlice con
 
 Result RocksDBCollection::read(transaction::Methods* trx,
                                arangodb::velocypack::StringRef const& key,
-                               ManagedDocumentResult& result, bool /*lock*/) {
```
Comment (joerg84) [inline arangod/RocksDBEngine/RocksDBCollection.cpp]: Is that the lock for the mmfile compatibilty workaround or just an api change?

File: arangod/StorageEngine/HotBackup.cpp [hunk 2]
```diff
@@ -46,7 +47,7 @@ HotBackup::HotBackup(application_features::ApplicationServer& server)
   } else if (EngineSelectorFeature::isRocksDB()) {
     _engine = BACKUP_ENGINE::ROCKSDB;
   } else {
-    _engine = BACKUP_ENGINE::MMFILES;
+    THROW_ARANGO_EXCEPTION_MESSAGE(TRI_ERROR_NOT_IMPLEMENTED, "mmfiles storage engine not available");
```
Comment (joerg84) [inline arangod/StorageEngine/HotBackup.cpp]: Maybe. same message as below `"hot backup not implemented for this storage engine"`

File: arangod/ClusterEngine/ClusterEngine.cpp [hunk 3]
```diff
@@ -108,9 +94,7 @@ ClusterEngineType ClusterEngine::engineType() const {
   }
   TRI_ASSERT(_actualEngine != nullptr);
```
[the remark this answers is not in the document]
Comment (jsteemann) [reply arangod/ClusterEngine/ClusterEngine.cpp:95]: `ClusterEngine::isMock()` also returns true when `_actualEngine` is a nullptr, but `ClusterEngine::Mocking` is set to true.
Some in `ClusterEngine::isMock()` we cannot rely on `_actualEngine` being a non-nullptr.

And both `isMock()` and `isRocksDB()` access the `name()` method of `_actualEngine` to check which engine is in use, but both only do this if `_actualEngine` is a non-nullptr. So it should be ok as it is.

File: arangod/ClusterEngine/ClusterEngine.cpp [hunk 4]
```diff
@@ -108,9 +94,7 @@ ClusterEngineType ClusterEngine::engineType() const {
...
 
-  if (isMMFiles()) {
-    return ClusterEngineType::MMFilesEngine;
-  } else if (isRocksDB()) {
+  if (isRocksDB()) {
```
[the remark this answers is not in the document]
Comment (jsteemann) [reply arangod/ClusterEngine/ClusterEngine.cpp:97]: Changed it to
```diff
@@ -94,12 +94,8 @@ ClusterEngineType ClusterEngine::engineType() const {
   }
   TRI_ASSERT(_actualEngine != nullptr);
 
-  if (isRocksDB()) {
-    return ClusterEngineType::RocksDBEngine;
-  }
-
-  TRI_ASSERT(false);
-  THROW_ARANGO_EXCEPTION_MESSAGE(TRI_ERROR_INTERNAL, "invalid engine type");
+  TRI_ASSERT(isRocksDB());
+  return ClusterEngineType::RocksDBEngine;
 }
 
```

File: arangod/ClusterEngine/ClusterRestCollectionHandler.cpp [hunk 5]
```diff
@@ -43,14 +42,6 @@ Result ClusterRestCollectionHandler::handleExtraCommandPut(LogicalCollection& co
       builder.add("result", VPackValue(true));
     }
     return res;
-  } else if (suffix == "rotate") {
```
[the remark this answers is not in the document]
Comment (jsteemann) [reply arangod/ClusterEngine/ClusterRestCollectionHandler.cpp:46]: The "rotate" command was previously only available for MMFiles. Here is its code:
```
43 int rotateActiveJournalOnAllDBServers(std::string const& dbname, std::string const& collname) {
 44   ClusterEngine* ce = static_cast<ClusterEngine*>(EngineSelectorFeature::ENGINE);
 45   if (!ce->isMMFiles()) {
 46     return TRI_ERROR_NOT_IMPLEMENTED;
 47   }
``` 
So it always returned `TRI_ERROR_NOT_IMPLEMENTED` for RocksDB, which after the change the method still does. So the user-facing behavior should not have changed.

File: arangod/StorageEngine/HotBackup.cpp (hunk 2 above)
Comment (jsteemann) [reply arangod/StorageEngine/HotBackup.cpp:50]: :+1:

**Comment by jsteemann** [comment]
http://172.16.10.101:8080/view/PR/job/arangodb-matrix-pr/9246/

File: arangod/ClusterEngine/ClusterRestCollectionHandler.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (joerg84) [reply arangod/ClusterEngine/ClusterRestCollectionHandler.cpp:46]: Thanks!

File: arangod/ClusterEngine/ClusterCollection.cpp [hunk 6]
```diff
@@ -64,31 +63,7 @@ ClusterCollection::ClusterCollection(LogicalCollection& collection, ClusterEngin
...
-    }
-
-    if (journalSlice.isNumber()) {
-      if (journalSlice.getNumericValue<uint32_t>() < TRI_JOURNAL_MINIMAL_SIZE) {
-        THROW_ARANGO_EXCEPTION_MESSAGE(TRI_ERROR_BAD_PARAMETER,
-                                       "<properties>.journalSize too small");
-      }
-    }
-  } else if (_engineType == ClusterEngineType::RocksDBEngine) {
+  if (_engineType == ClusterEngineType::RocksDBEngine) {
```
Comment (gnusi) [inline arangod/ClusterEngine/ClusterCollection.cpp:66]: shouldn't we use `assert` instead?

File: arangod/ClusterEngine/ClusterCollection.cpp [hunk 7]
```diff
@@ -139,52 +114,7 @@ Result ClusterCollection::updateProperties(VPackSlice const& slice, bool doSync)
...
-
-    merge.add("doCompact",
-              VPackValue(Helper::getBooleanValue(slice, "doCompact", true)));
-    merge.add("indexBuckets",
-              VPackValue(Helper::getNumericValue(slice, "indexBuckets",
-                                                  MMFilesCollection::defaultIndexBuckets)));
-    merge.add("journalSize", VPackValue(journalSize));
-
-  } else if (_engineType == ClusterEngineType::RocksDBEngine) {
+  if (_engineType == ClusterEngineType::RocksDBEngine) {
```
Comment (gnusi) [inline arangod/ClusterEngine/ClusterCollection.cpp:117]: why not just `assert`?

File: arangod/RestServer/DatabaseFeature.cpp [hunk 8]
```diff
@@ -574,9 +574,6 @@ void DatabaseFeature::recoveryDone() {
       continue;
     }
 
-    // execute the engine-specific callbacks on successful recovery
-    engine->recoveryDone(*vocbase);
```
Comment (gnusi) [inline arangod/RestServer/DatabaseFeature.cpp:578]: I bet that may break ArangoSearch recovery logic

File: arangod/RestServer/DatabaseFeature.cpp [hunk 9]
```diff
@@ -686,9 +683,6 @@ Result DatabaseFeature::createDatabase(CreateDatabaseInfo&& info, TRI_vocbase_t*
     }
 
     if (!engine->inRecovery()) {
-      // starts compactor etc.
-      engine->recoveryDone(*vocbase);
```
Comment (gnusi) [inline arangod/RestServer/DatabaseFeature.cpp:690]: I bet that may break ArangoSearch recovery logic

File: arangod/VocBase/Methods/UpgradeTasks.cpp [hunk 10]
```diff
@@ -545,24 +544,8 @@ bool UpgradeTasks::addDefaultUserOther(TRI_vocbase_t& vocbase,
   return true;
 }
 
-bool UpgradeTasks::persistLocalDocumentIds(TRI_vocbase_t& vocbase,
-                                           arangodb::velocypack::Slice const& slice) {
-  if (EngineSelectorFeature::engineName() != MMFilesEngine::EngineName) {
```
Comment (gnusi) [inline arangod/VocBase/Methods/UpgradeTasks.cpp:550]: I'd prefer to leave at least assertion for the future

File: arangod/VocBase/Methods/UpgradeTasks.cpp [hunk 11]
```diff
@@ -545,24 +544,8 @@ bool UpgradeTasks::addDefaultUserOther(TRI_vocbase_t& vocbase,
...
-  Result res = basics::catchToResult([&vocbase]() -> Result {
-    MMFilesEngine* engine = static_cast<MMFilesEngine*>(EngineSelectorFeature::ENGINE);
-    return engine->persistLocalDocumentIds(vocbase);
-  });
-  return res.ok();
-}
-
 bool UpgradeTasks::renameReplicationApplierStateFiles(TRI_vocbase_t& vocbase,
                                                       arangodb::velocypack::Slice const& slice) {
-  if (EngineSelectorFeature::engineName() == MMFilesEngine::EngineName) {
```
Comment (gnusi) [inline arangod/VocBase/Methods/UpgradeTasks.cpp:562]: I'd prefer to leave at least assertion for the future

File: arangod/RestServer/FlushFeature.cpp [hunk 12]
```diff
@@ -129,8 +124,6 @@ arangodb::Result FlushFeature::releaseUnusedTicks(size_t& count, TRI_voc_tick_t&
     TRI_TerminateDebugging("crashing before syncing min tick");
   }
 
-  engine->waitForSyncTick(minTick);
```
Comment (gnusi) [inline arangod/RestServer/FlushFeature.cpp:132]: I'd prefer to leave a comment that WAL tick has to be synced prior releasing it, again for the future

File: arangod/ClusterEngine/ClusterCollection.cpp (hunk 6 above)
Comment (jsteemann) [reply arangod/ClusterEngine/ClusterCollection.cpp:66]: There is still the MockEngine, so we need at least one `if`.

File: arangod/ClusterEngine/ClusterCollection.cpp (hunk 7 above)
Comment (jsteemann) [reply arangod/ClusterEngine/ClusterCollection.cpp:117]: There is still the MockEngine, so we need at least one `if`.

File: arangod/RestServer/DatabaseFeature.cpp (hunk 8 above)
Comment (jsteemann) [reply arangod/RestServer/DatabaseFeature.cpp:578]: I don't think so. `recoveryDone(vocbase)` was a no-op in all storage engines but MMFiles.

File: arangod/RestServer/DatabaseFeature.cpp (hunk 9 above)
Comment (jsteemann) [reply arangod/RestServer/DatabaseFeature.cpp:690]: I don't think so. `recoveryDone(vocbase)` was a no-op in all storage engines but MMFiles.

File: arangod/VocBase/Methods/UpgradeTasks.cpp (hunk 10 above)
Comment (jsteemann) [reply arangod/VocBase/Methods/UpgradeTasks.cpp:550]: Mmmh, the entire method has been deleted. Where to put the assertion?

File: arangod/VocBase/Methods/UpgradeTasks.cpp (hunk 11 above)
Comment (jsteemann) [reply arangod/VocBase/Methods/UpgradeTasks.cpp:562]: Ok, added the assertion.

File: arangod/RestServer/FlushFeature.cpp (hunk 12 above)
Comment (jsteemann) [reply arangod/RestServer/FlushFeature.cpp:132]: Ok, did!

**Comment by jsteemann** [comment]
http://172.16.10.101:8080/view/PR/job/arangodb-matrix-pr/9250/

File: arangod/RestServer/DatabaseFeature.cpp (hunk 8 above)
Comment (gnusi) [reply arangod/RestServer/DatabaseFeature.cpp:578]: you're right, I messed it up with `DatabaseFeature::recoveryDone()` itself :-)

File: arangod/VocBase/Methods/UpgradeTasks.cpp (hunk 10 above)
Comment (gnusi) [reply arangod/VocBase/Methods/UpgradeTasks.cpp:550]: :-)

**Comment by jsteemann** [comment]
http://172.16.10.101:8080/view/PR/job/arangodb-matrix-pr/9277/

**Pull request merged**